src/Compilers/VisualBasic/Test/Semantic/Diagnostics/DiagnosticAnalyzerTests.vb

```diff
@@ -812,5 +813,240 @@ End Class
+
+            ' Expected compilation diagnostics
+            AddExpectedNonLocalDiagnostic(builder, "PartialType", compilation.SyntaxTrees(0).FilePath)
+
+            Dim expected = builder.ToArrayAndFree()
+
+            VerifyGeneratedCodeAnalyzerDiagnostics(compilation, expected, generatedCodeAnalysisFlagsOpt:=Nothing)
+            VerifyGeneratedCodeAnalyzerDiagnostics(compilation, expected, GeneratedCodeAnalysisFlags.None)
+            VerifyGeneratedCodeAnalyzerDiagnostics(compilation, expected, GeneratedCodeAnalysisFlags.Default)
+            VerifyGeneratedCodeAnalyzerDiagnostics(compilation, expected, GeneratedCodeAnalysisFlags.Analyze)
+            VerifyGeneratedCodeAnalyzerDiagnostics(compilation, expected, GeneratedCodeAnalysisFlags.ReportDiagnostics)
+            VerifyGeneratedCodeAnalyzerDiagnostics(compilation, expected, GeneratedCodeAnalysisFlags.Analyze Or GeneratedCodeAnalysisFlags.ReportDiagnostics)
+        End Sub
+
+        Private Shared Sub VerifyGeneratedCodeAnalyzerDiagnostics(compilation As Compilation, isGeneratedFileName As Func(Of String, Boolean), generatedCodeAnalysisFlagsOpt As GeneratedCodeAnalysisFlags?)
+            Dim expected = GetExpectedGeneratedCodeAnalyzerDiagnostics(compilation, isGeneratedFileName, generatedCodeAnalysisFlagsOpt)
+            VerifyGeneratedCodeAnalyzerDiagnostics(compilation, expected, generatedCodeAnalysisFlagsOpt)
+        End Sub
+
+        Private Shared Sub VerifyGeneratedCodeAnalyzerDiagnostics(compilation As Compilation, expected As DiagnosticDescription(), generatedCodeAnalysisFlagsOpt As GeneratedCodeAnalysisFlags?)
+            Dim analyzers = New DiagnosticAnalyzer() {New GeneratedCodeAnalyzer(generatedCodeAnalysisFlagsOpt)}
+            compilation.VerifyAnalyzerDiagnostics(analyzers, Nothing, Nothing, False, expected)
+        End Sub
+
+        Private Shared Function GetExpectedGeneratedCodeAnalyzerDiagnostics(compilation As Compilation, isGeneratedFileName As Func(Of String, Boolean), generatedCodeAnalysisFlagsOpt As GeneratedCodeAnalysisFlags?) As DiagnosticDescription()
+            Dim analyzers = New DiagnosticAnalyzer() {New GeneratedCodeAnalyzer(generatedCodeAnalysisFlagsOpt)}
+            Dim files = compilation.SyntaxTrees.Select(Function(t) t.FilePath).ToImmutableArray()
+            Dim sortedCallbackSymbolNames = New SortedSet(Of String)()
+            Dim sortedCallbackTreePaths = New SortedSet(Of String)()
+            Dim builder = ArrayBuilder(Of DiagnosticDescription).GetInstance()
+            For i As Integer = 1 To 9
+                Dim file = files(i - 1)
+                Dim isGeneratedFile = isGeneratedFileName(file)
+
+                ' Type "GeneratedCode{0}"
+                Dim squiggledText = String.Format("GeneratedCode{0}", i)
+                Dim diagnosticArgument = squiggledText
+                Dim line = If(i < 9, 3, 4)
```

Comment:
I have cleaned up the tests a bit, getting rid of magic numbers.

This test is basically producing a bunch of source files with different combinations (generated vs non-generated extension, auto generated comments, GeneratedCodeAttribute, etc.). The underlying analyzer being executed generates diagnostics for symbols and syntax trees it analyzes and also tracks what entities it received analyzer callbacks for to report a summary diagnostic in compilation end action.